### PR TITLE
updated agent according to new action space, experience replay disabled

### PR DESCRIPTION
**Pull request by zhouzypaul**
pls review the TODO comments in `agent.py`. I might have messed things up.

### CODE REVIEW

File: agent.py [hunk 1]
```diff
@@ -91,24 +94,36 @@ def call(self, states):
...
+                particular stock on a particular day sums to one)
                 a tensor that summarize the states tensor of shape [batch_sz * self.actor_H2]
         """
-        gru_1_out_whole_seq, _ = self.actor_dropout_1(self.actor_gru_1(states))  # [batch_sz * past_num * actor_H1]
-        gru_2_out = self.actor_dropout_2(self.actor_gru_2(gru_1_out_whole_seq))  # [batch_sz * actor_H2]
-        actor_out = self.actor_dense(gru_2_out)  # [batch_sz * self.num_actions]
-        return actor_out, gru_2_out
+        # extract info from states
+        price_history = [state[0] for state in states]  # [batch_sz, num_stocks, past_num, datum_size]
+        price_history = tf.reshape(price_history, (-1, self.num_stocks * self.past_num, self.datum_size))  # TODO: this seems wrong
```
Comment (seanxzhan) [inline agent.py]: i checked this on jupyter notebook (made a tensor and reshaped it this way), this seems to work.

File: agent.py [hunk 2]
```diff
@@ -91,24 +94,36 @@ def call(self, states):
...
+        portfolio = [state[1] for state in states]  # [batch_sz * (num_stock+1)]
+        # pass through layers
+        gru_1_out_whole_seq, _ = self.actor_dropout_1(self.actor_gru_1(price_history))  # (batch_sz, num_stock * past_num , actor_H1)
+        gru_2_out = self.actor_dropout_2(self.actor_gru_2(gru_1_out_whole_seq))  # [batch_sz, actor_H2]
+        past_and_current_info = tf.concat([gru_2_out, portfolio], axis=1)  # (batch_sz, actor_H2 + num_stock + 1)
+        actor_out = self.actor_dense(past_and_current_info)  # (batch_sz * self.num_actions)
+        # reshape the output and softmax
+        actor_out = tf.reshape(actor_out, (-1, self.num_stocks, 3))
+        action_probs = tf.nn.softmax(actor_out, axis=-1)  # TODO: not sure if this sums the last dim to 1
```
Comment (seanxzhan) [inline agent.py]: this is correct

File: agent.py (hunk 1 above)
Comment (zhouzypaul) [reply agent.py]: my main concern is whether the 2nd dim is justified to be of size `num_stocks * past_num`, because this would then become the `timestep` param of the GRU, which we may want to be just `past_num` (?)

sorry if my initial comments were not clear

File: agent.py [hunk 3]
```diff
@@ -140,19 +140,31 @@ def loss(self, states, actions_taken, discounted_reward):
...
+#                                                      list(zip(np.arange(self.num_stocks), actions_taken[b]))))
+#             # TODO: this could be very wrong...
+#         Above has error: 'tensorflow.python.framework.ops.EagerTensor' object does not support item assignment
+#         you could get around this with setattr, but not sure if it's gonna mess with backprop. anyways, below 
+#         might be a little more efficient (no for loop):
+        hot_tens = tf.one_hot(actions_taken, 3) # depth of 3, since we have 3 possible actions
+        # only desired indices will be multiplied by 1, others are multiplied by 0
+        filtered_tens = tf.multiply(action_probs, hot_tens) 
+        # add the last dimension to get rid of 0's from filtered_tens
+        probs_action_taken_each_stock = tf.keras.backend.sum(filtered_tens, axis=-1) # (batch_sz, num_stocks)
```
Comment (zhouzypaul) [inline agent.py:165]: is `tf.keras.backend.sum` different than `tf.reduce_sum`? just wondering, verified that it works.

Comment (seanxzhan) [reply agent.py:165]: short answer: no they are not much different. long answer: tf.keras.backend.sum is similar to numpy.sum() which allows us to simply sum across a dimension without reducing. I just looked at tf.reduce_sum again and realized I could have set tf.reduce_sum(..., keepdims=True) to prevent reducing, but I guess we can j keep it as is for now.

File: agent.py (hunk 1 above)
Comment (seanzhan0319) [reply agent.py:107]: haha you are good

you know how we passed embedding_size as the timestep parameter for the language models assigment? I think num_stocks * past_num would accurately represent the "embedding" here because in this way the second dimension reflects a snapshot of our entire portfolio in the past number of days

Comment (AdamWang00) [reply agent.py]: fixed - 66316bf
